src/FSharp.Management.PowerShell/PowerShellProvider.fs

```diff
@@ -52,22 +50,54 @@ type public PowerShellProvider(_cfg: TypeProviderConfig) as this =
+                                            failwithf "Expected %d arguments but received %d" paramList.Length args.Length
 
-                                    let elements = [ for i in 0..(paramList.Length-1) ->
-                                                        Quotations.Expr.Coerce(args.[i], typeof<obj>) ]
-                                    let parameters = Quotations.Expr.NewArray(typeof<obj>, elements)
+                                        let elements = [ for i in 0..(paramList.Length-1) ->
+                                                            Quotations.Expr.Coerce(args.[i], typeof<obj>) ]
+                                        Quotations.Expr.NewArray(typeof<obj>, elements)
+                                        
                                     let uid = cmdlet.UniqueID
-                                    <@@ Runtime.Current(psSnapIns.Split(';'), modules.Split(';'), is64bitRequired, false)
-                                                .Execute(uid, (%%parameters : obj[])) @@>)
+                                    if areStatic then 
+                                        let parameters = mapToParameters args
+                                        <@@ Runtime.Current(psSnapIns.Split(';'), modules.Split(';'), is64bitRequired, false)
+                                                .Execute(uid, (%%parameters : obj[])) @@>
+                                    else 
+                                        let constructorExpression, args = 
+                                            match args with
+                                            | head :: tail -> head, tail
+                                            | [] -> failwith "Constructor should create runspace"
+
+                                        let runspace = Quotations.Expr.Coerce(constructorExpression, typeof<HostedRuntime.IPSRuntime>) 
+                                        let parameters = mapToParameters args
+                                        <@@ (%%runspace : HostedRuntime.IPSRuntime).Execute(uid, (%%parameters : obj[])) @@>)
+                                    
+
                     providedMethod.AddXmlDocDelayed(fun() ->runtime.GetXmlDoc(cmdlet.Name))
                     yield providedMethod :> MemberInfo
-                ])
+                ]
+
+            let pty = ProvidedTypeDefinition(thisAssembly, rootNamespace, typeName, Some(baseTy))
+            pty.AddMembersDelayed(fun() -> createMembers true)
+
+            let customRunspace = ProvidedTypeDefinition("CustomRunspace", Some(typeof<obj>))
+            customRunspace.AddMember <| ProvidedConstructor([], InvokeCode = fun args -> <@@ new HostedRuntime.PSRuntimeHosted(psSnapIns.Split(';'), modules.Split(';')) :> HostedRuntime.IPSRuntime @@>)
+            customRunspace.AddInterfaceImplementation <| typeof<System.IDisposable>
+            customRunspace.AddMember <| ProvidedMethod("Dispose", [], typeof<unit>, 
```

Comment:
I a, really not sure that this is correct interface implementation.
Could you please test that when you do `use runspace = new PS.CustomRunspace()` is actually call `Dispose()` method?

I think that when you use `AddMember` you hide `Dispose` impl, so `(runtime :> IDisposable).Dispose()` will not call this method. I guess that we should use `customRunspace.DefineMethodOverride` to provide `Dispose` impl.

Here is a sample from `FsXaml` - https://github.com/fsprojects/FsXaml/blob/c0979473eddf424f7df83e1b9222a8ca9707c45a/src/FsXaml.Wpf.TypeProvider/XamlTypeProvider.fs#L100-L102